src/core/rotation.cpp

```diff
@@ -550,4 +567,269 @@ void rotate_particle(Particle *p, double *aSpaceFrame, double phi) {
 #endif
 }
 
+/** Rotate the particle p around the body axis "a" by amount phi */
+void rotate_particle_body(Particle* p, double* a, double phi)
+{
+  // Apply restrictions from the rotation_per_particle feature
+#ifdef ROTATION_PER_PARTICLE
+  // Rotation turned off entirely?
+  if (p->p.rotation <2) return;
+
+  // Per coordinate fixing
+  if (!(p->p.rotation & 2)) a[0]=0;
+  if (!(p->p.rotation & 4)) a[1]=0;
+  if (!(p->p.rotation & 8)) a[2]=0;
+  // Re-normalize rotation axis
+  double l=sqrt(sqrlen(a));
+  // Check, if the rotation axis is nonzero
+  if (l<1E-10) return;
+
+  for (int i=0;i<3;i++)
+    a[i]/=l;
+
+#endif
+
+  double q[4];
+  q[0]=cos(phi/2);
+  double tmp=sin(phi/2);
+  q[1]=tmp*a[0];
+  q[2]=tmp*a[1];
+  q[3]=tmp*a[2];
```

Comment:
This unreadable version has zero performance advantage in the presence of optimizations.  Just use
```cpp
double q[] = {
    cos(phi/2),
    sin(phi/2)*a[0],
    sin(phi/2)*a[1],
    sin(phi/2)*a[2]
};
```
This produces identical assembly, see [Compiler Explorer](https://godbolt.org/g/asJvpS).

Reply:
Agree / done.